### DISCUSSION TITLE
Stochastic Optimizers without control parameter

### DISCUSSION THREAD

**Discussion by hwborchers** [SudoWodo/StochOptim#10] [General]
I am in doubt about the naming conventions of StochOptim at the moment. Imagine I want to add my own solver `simpleDE` in package 'adagio' -- and there are two of them, so I need to distinguish them. I added the following to `LocalWrapper.R`, BUT: the naming of functions in StochOptim is getting nasty now.

```
localWrapper_adagio_simpleDE <-
    function(fn, lower, upper, method, control = list()) {
    
    # this calls the constructor and new object is formed
    obj <- adagio_simpleDEWrapper$new(
        fn      = fn, 
        lower   = lower, 
        upper   = upper,
        method  = method,
        control = control
    )
    
    # package installation check
    obj$checkinstallation()
    
    # check controls
    obj$checkcontrol()
    
    # object calls the optimizer
    ans <- obj$callOptimizer()
    
    # printout the answer in std. format
    obj$printoutput()
    
    return(ans)
}
```

In the `OptimizerWrapper.R` file I added the following, BUT: `simpleDE` does not have a `control` parameter. How are we going to handle this?

```
adagio_simpleDEWrapper <- R6Class(
    
    classname = "adagio_simpleDEWrapper",
    inherit = OptimizerWrapper,
    public = list(
        
        vcontrol = c("N", "nmax", "r", "confined", "log"),
        
        # call the optimizer 
        callOptimizer = function(){
            self$ans <- adagio::simpleDE(fn      = self$fn, 
                                         lower   = self$lower, 
                                         upper   = self$upper, 
                                         control = self$control
            )
            
            return(self$ans)
        },
        
        # for nicely printing out the answer
        printoutput = function(){
            output <- list(
                par     = self$ans$xmin,
                value   = self$ans$fmin,
                counts  = c(`function` = self$ans$nfeval)
            )
            
            print(output)
        }
        
    ) # end public list
) # end adagio_simpleDEWrapper class
```

**Comment by SudoWodo** [comment]
Sir, it seems like you are trying to add the new solver in the prototype and which does NOT contain the right naming scheme and seeing how naming is going now even i feel unsure.

## Naming scheme
So according to the naming policy that i have suggest for the package your class name would be `adagio_simpleDE_wrapper` and the local wrapper name would be `localWrapperadagiosimpleDE`. **To be honest I am feeling odd about this local wrapper name.**

## Control parameter and standardization

I believe the real issue with `adagio::simpleDE ` not having the control list is standardization. I am certain that the class wrapper can be molded to accommodate this issue.  But still there are many things to be discussed and decided here.


## The proposed solution
The first sensible way that comes to my mind is have a control list just as we make for other optimizer then let the wrappers handle the translation/mapping of these parameters to the optimizer. My attempt to implement this is as follows:-  
1. Check the passed control list against vcontrol
2. Define default_control variable in the class which will hold default values as defined in the package and method
3. Change the default_control with the passed control
4. Call the optimizer


**Class definition**
```
adagio_simpleDE_wrapper <- R6Class(
  
  classname = "adagio_simpleDE_wrapper",
  inherit = OptimizerWrapper,
  public = list(
    
    vcontrol = c("N", "nmax", "r", "confined", "log"),
    
    # set default control
    default_control = list(N        = 64, 
                           nmax     = 256,
                           r        = 0.4, 
                           confined = TRUE, 
                           log      = FALSE),
    
    # change default_control with passsed control
    changedefaultcontrol = function() {
      for (i in names(self$control)) {
        self$default_control[i] <- self$control[i]
      }
      self$control <- self$default_control
    }
    ,
    
    # call the optimizer 
    callOptimizer = function() {
      self$ans <- adagio::simpleDE(fun      = self$fn,
                                   lower    = self$lower,
                                   upper    = self$upper,
                                   N        = self$control$N,
                                   nmax     = self$control$nmax,
                                   r        = self$control$r,
                                   confined = self$control$confined,
                                   log      = self$control$log)
      
      return(self$ans)
    },
    
    # for nicely printing out the answer
    printoutput = function() {
      output <- list(
        par     = self$ans$xmin,
        value   = self$ans$fmin,
        counts  = c(`function` = self$ans$nfeval)
      )
      
      print(output)
    }
    
  ) # end public list
) # end adagio_simpleDE_wrapper class

```

**Local wrapper definition**

`localWrapperadagiosimpleDE` seems like a bad name so I'll not use it.

```
local_wrapper_adagio_simpleDE <-
  function(fn, lower, upper, method, control = list()) {
    
    # this calls the constructor and new object is formed
    obj <- adagio_simpleDE_wrapper$new(
      fn      = fn, 
      lower   = lower, 
      upper   = upper,
      method  = method,
      control = control
    )
    
    
    # package installation check
    obj$checkinstallation()
    
    # check controls
    obj$checkcontrol()
    
    # change default control
    obj$changedefaultcontrol()
    
    # object calls the optimizer
    ans <- obj$callOptimizer()
    
    # printout the answer in std. format
    obj$printoutput()
    
    return(ans)
  }
```

**example**

```
rastrigin <- function(x) {
  y <- ( 10 * length(x) + sum(x^2 - 10 * cos(2 * pi * x)))
  return(y)
}

D <- 10
lb <- rep(-5.2, D)
ub <- rep(5.2, D)
ctrl5 <- list(N        = 70, 
              nmax     = 230,
              r        = 0.41, 
              confined = TRUE, 
              log      = FALSE)

local_wrapper_adagio_simpleDE(fn = rastrigin, lb, ub, method = "adagio", control = ctrl5)
```

**Comment by hwborchers** [comment]
The global wrapper will still use "population size" as an option. So we need a connection between population size and the parameter `N` of the `simpleDE` function. Where is this done? - sorry I can't find this connection at the moment.

**Reply by SudoWodo** [reply]
Actually there is no such connection yet. I have been wanting to ask this question and understand the parameter that need to be present in global wrapper for example "population size" you just mentioned. Apart from population size what are some other parameters that need to be there? are we going to make them formal parameters or items in the standard control list?

**Reply by hwborchers** [reply]
Some time ago I wrote some lines for a prototype of a global optimization package. Here is an example of something we now would call a "local wrapper":

```r
if (method == "deoptim") {
        sol <- DEoptim::DEoptim(fn, lower = lb, upper = ub,
                                DEoptim::DEoptim.control(
                                trace = cntrl$info, itermax = cntrl$maxiter))
        return(list(xmin = sol$optim$bestmem, fmin = sol$optim$bestval,
                    niter = sol$optim$iter, nfeval = sol$optim$nfeval,
                    commment = ""))
```

resp.

```r
    } else if (method == "simplede") {
        sol <- adagio::simpleDE(fun = fn, lower = lb, upper = ub,
                                N = cntrl$popsize, nmax = cntrl$maxiter,
                                log = cntrl$info)
        return(list(xmin = sol$xmin, fmin = sol$fmin))
```

The local wrapper does two things:

- it transforms global options such as population size max. iterations, etc., into parameters or control options of the stochastic solver, and

- it transforms the components of the result list into a formalized list such that all returns from local wrappers have the same structure.

This is what I would expect from a local wrapper. I admit I have no clear idea how to integrate this into an object-oriented approach. The other mentors may have different proposals. I will be glad to hear from them.

**Reply by hwborchers** [reply]
The question which options the global wrapper will accept, that is options that all stochastic optimizers understand, has not really been answered. In my understanding, the following are possible global options:

- initial point (some global optimizers need it)
- lower and upper bounds
- maximum number of iterations
- size of population
- tolerance (used as relative tolerance in some optimizers)

We need to go through more stochastic optimizers to identify such global options. Options that have no default values are among this class.

Then there are specific control options that are used by one or only very few optimizers. For example, `strategy` is only known for 'DEoptim' and 'DEoptimR' I think. These options probably have to be made known to the optimizer through some other mechanism.

**Reply by nashjc** [reply]
These overlap:
- initial point (some global optimizers need it)
- lower and upper bounds
One can specify bounds in some solvers, in which case "initial" point or points are computed e.g., mean of bounds.
In some the bounds are NOT given, but a stepsize is given that creates a zone for the search.

**Reply by nashjc** [reply]
Note that "iterations" is a very slippery concept. In some methods a single iteration does a lot of work; in others it may involve only a single function evaluation. I think function (and if relevant gradient and hessian) evaluation counts are useful, as are counts of major operations, such as the solution of Newton equations or similar sets of equations.
